### PR TITLE
Internal refactoring of source locators

### LINKED ISSUE #123: SourceLocators should return *definitely* located source, not *potentially* located source
Currently the `SourceLocators` return "potentially" located source, which is causing issues such as described in #118 (inability to use multiple `StringSourceLocator`s for example). The fact it is unclear if a located source actually contains the requested identifier is pretty weak and flaky behaviour.

A `SourceLocator` should return "definitely" located source, maybe by peeking at the AST to ensure the requested identifier can be located. Need to be careful to re-use/abstract the logic contained in `Reflector\Generic` so that we don't dupe code and become unperformant.

### PR DESCRIPTION
**Pull request by asgrim**
Fixes #123 and #43

* Organised the `SourceLocator` folder nicer
* `Reflector\Generic` is now gone
   * Refactored into the `SourceLocator\Ast\Locator` and related classes
* Refactored all `SourceLocators` to use the `SourceLocator\Ast\Locator` that "definitely" returns source

### CODE REVIEW

**Comment by asgrim** [comment]
@asgrim needs moar tests

**Comment by asgrim** [comment]
@Ocramius tests added, coverage is 98.97%

File: src/SourceLocator/Ast/FindReflectionsInTree.php [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+namespace BetterReflection\SourceLocator\Ast;
+
+use BetterReflection\Identifier\IdentifierType;
+use BetterReflection\Reflector\Reflector;
+use BetterReflection\SourceLocator\Ast\Strategy\NodeToReflection;
+use BetterReflection\SourceLocator\Located\LocatedSource;
+use BetterReflection\Reflection\Reflection;
+use PhpParser\Node;
+
+class FindReflectionsInTree
```
Comment (Ocramius) [inline src/SourceLocator/Ast/FindReflectionsInTree.php]: `final`

File: src/SourceLocator/Ast/Strategy/NodeToReflection.php [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+use BetterReflection\SourceLocator\Located\LocatedSource;
+use BetterReflection\Reflection\ReflectionClass;
+use BetterReflection\Reflection\ReflectionFunction;
+use BetterReflection\Reflection\Reflection;
+use PhpParser\Node;
+
+/**
+ * @internal
+ */
+class NodeToReflection
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Strategy/NodeToReflection.php]: `final`

File: src/SourceLocator/Type/AbstractSourceLocator.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+use BetterReflection\Identifier\IdentifierType;
+use BetterReflection\Reflector\Exception\IdentifierNotFound;
+use BetterReflection\Reflector\Reflector;
+use BetterReflection\SourceLocator\Ast\Locator as AstLocator;
+use BetterReflection\SourceLocator\Located\LocatedSource;
+
+/**
+ * Use PHP's built in autoloader to locate a class, without actually loading.
+ *
+ * There are some prerequisites...
```
Comment (Ocramius) [inline src/SourceLocator/Type/AbstractSourceLocator.php]: This should probably not be on the abstract source locator

File: src/SourceLocator/Type/AbstractSourceLocator.php [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+use BetterReflection\Identifier\Identifier;
+use BetterReflection\Identifier\IdentifierType;
+use BetterReflection\Reflector\Exception\IdentifierNotFound;
+use BetterReflection\Reflector\Reflector;
+use BetterReflection\SourceLocator\Ast\Locator as AstLocator;
+use BetterReflection\SourceLocator\Located\LocatedSource;
+
+/**
+ * Use PHP's built in autoloader to locate a class, without actually loading.
```
Comment (Ocramius) [inline src/SourceLocator/Type/AbstractSourceLocator.php]: Probably wrong description

File: src/SourceLocator/Ast/Locator.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+use BetterReflection\Identifier\IdentifierType;
+use BetterReflection\SourceLocator\Ast\Strategy\NodeToReflection;
+use BetterReflection\Reflection\Reflection;
+use BetterReflection\Reflector\Reflector;
+use BetterReflection\Reflector\Exception\IdentifierNotFound;
+use BetterReflection\SourceLocator\Located\LocatedSource;
+use PhpParser\Parser;
+use PhpParser\Lexer;
+
+class Locator
```
Comment (Ocramius) [inline src/SourceLocator/Ast/Locator.php]: Needs an interface?

File: test/unit/Reflection/ReflectionObjectTest.php [hunk 6]
```diff
@@ -31,7 +31,7 @@ private function getPhpParser()
 
         $this->parser = new Parser\Multiple([
```
Comment (Ocramius) [inline test/unit/Reflection/ReflectionObjectTest.php]: We can probably use the parser factory here now

**Pull request merged**